sw/cheri/tests/pwm_tests.hh

```diff
@@ -0,0 +1,207 @@
+// Copyright lowRISC Contributors.
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+#include "../../common/defs.h"
+#include "../common/console.hh"
+#include "../common/block_tests.hh"
+#include "../common/timer-utils.hh"
+#include "test_runner.hh"
+#include <cheri.hh>
+#include <platform-gpio.hh>
+#include <platform-pwm.hh>
+
+using namespace CHERI;
+
+typedef SonataPulseWidthModulation SonataPwm;
+
+/**
+ * Configures the number of test iterations to perform.
+ * This can be overriden via a compilation flag.
+ */
+#ifndef PWM_TEST_ITERATIONS
+#define PWM_TEST_ITERATIONS (1U)
+#endif
+
+/**
+ * Configures whether cable connections required for PWM testing
+ * are available. This includes cables between:
+ *  - MikroBus PWM and PMOD0 Pin 1
+ * This can be overriden via a compilation flag.
+ */
+#ifndef PWM_CABLE_CONNECTIONS_AVAILABLE
+#define PWM_CABLE_CONNECTIONS_AVAILABLE true
+#endif
+
+// Test Parameters
+constexpr uint32_t NumPwmCyclesObserved  = 25;
+constexpr uint32_t AllowedCycleDeviation = 20;
+constexpr bool LogPwmTests               = false;
+
+/**
+ * Runs a PWM loopback tests, which uses a jumper cable connecting a PWM output pin to a GPIO input pin to perform
+ * measurements of the PWM output, and test whether the PWM is acting correctly as configured within a certain margin of
+ * error.
+ *
+ * Requires a connection between PWM 0 (mikroBus PWM) and PMOD0 pin 1.
+ *
+ * @param gpio_pmod0 Capability for PMOD0's GPIO
+ * @param pwm PWM driver
+ * @param period The PWM period to test with
+ * @param duty_cycle The PWM duty cycle to test with. Represented as a separate counter, so for example a duty cycle of
+ * 50 with a period of 200 would give 1/4 i.e. 25% high output.
+ * @param cycles The number of PWM cycles to test
+ * @param error_delta The margin of error on either side of the number of cycles measured for the period and duty cycle.
+ * There will always be some error, as measuring logic cannot run on every cycle.
+ *
+ * @returns The integer number of failures during the test
+ */
+int pwm_loopback_test(Capability<volatile SonataGpioPmod> gpio_pmod0, Capability<volatile SonataPwm> pwm,
```

Comment:
Isn't there aliases for these capabilities in `common/sonata-devices.hh`

Reply:
No, `SonataGpioPmod` is a part of the new GPIO driver and I don't want to make sweeping changes to use that here because it will likely soon be replaced by `SonataGpioPmod0` after the new pin mapping changes are merged. Likewise, `SonataPwm` hasn't been used before so I suppose I could add an instance of it to `sonata-devices.hh`, but again I think that is better addressed in a follow-up PR which cleans up some of these instances, because the new mapping changes will change the PWM driver to use 7 instances instead of just 1.